### ISSUE TITLE
what classes do we need and what should be in each?

### ISSUE THREAD

**Issue by jspaaks** [matchms/matchms-backup#41]
1. ``Spectrum``: _in-memory representation of a spectrum file_

    methods, properties:
       - ``def load(self)`` _loading spectrum data from a file_
       - methods for slicing the tabular self.data
       - self.filename _source of the data_
       - self.data _contents of the file_
       - others?

    I'm as yet undecided on whether the filtering and merging that is currently in the Spectrum class, should remain there. I suspect the filtering and merging is done in an attempt to harmonize the data so it may be more easily compared, or to speed up such a comparison.

1. a new class ``Performance``, ``Similarity``, ``Scores`` or something to that effect, somewhat similar in spirit to ``SimilarityMeasures`` from ``similarity_measure.py``. Central to this class is a table of similarity scores. Initially the table is empty.  A user can then add a similarity measure function such as _Tanimoto_ or _modified cosine_ or whatever, thereby creating a column in the table. A user can also add reference spectrums to the table, they become the rows. All the classical comparison methods can do 1-on-1 comparison, maybe this is something we could parallelize?. The interface for all the classical similarity measure functions should then be ``def score = (spectrum Spectrum, reference_spectrum Spectrum) -> float`` (excuse the mix of languages there). For the word2vec based similarity measure, I propose that the interface should take an additional argument ``model `` (if we can come up with a more disambiguous term, that might help). If we choose to organize it this way, we can cleanly separate the complete body of word2vec-related stuff from all the classical stuff, while retaining flexibilty to extend the ``Scores`` table with other similarity measure functions, either based on the classical interface, or on the new interface. Another advantage would be that all the word2vec analysis yields just one variable, ``model``.

    Maybe harmonization should happen in this class, as opposed to in ``Spectrum`` --that way, it's easier to ensure that all spectrums are subjected to the same harmonization routine. If we go this way, it's probably a good idea to add a method ``def add_harmonization_function(self, func)``.

    methods:
       - ``def add_similary_measure_function(self, func)``
       - ``def add_reference_spectrums(self, spectrums Spectrum[])``
       - ``def add_harmonization_function(self, func)`` _probably, we should extend the method so it can take an index, so users have control over the order in which the harmonizations are run (assuming that matters)_
       - ``def calculate(self)`` _take a spectrum, run the harmonization functions on it in the right order, then take a reference_spectrum, and run the harmonization functions on it in the right order, then feed the two harmonized spectrums to the first similarity_measure_function, then the second, etc. and fill in the corresponding row of the table. Iterate over all elements of the table. It's probably a good idea to offer functionality to only calculate certain parts of the table, for example only a subset of similarity measure functions, or only a subset of the reference spectrums. If we go this way, you could also implement ``__add__``, which would allow users to concatenate two or more sections of the same ``Scores`` table. With that ability, you can let ``calculate`` divide the tasks over multiple machines for quicker evaluation_

    properties
       - self.spectrum
       - self.reference_spectrums
       - self.similarity_measure_functions
       - self.harmonization_functions

    The instance variable of ``Scores`` should then be used as input for any kind of search, plotting, and writing of output.

1. others? we have ``EpochLogger`` from ``similarity_measures.py`` left over, or we could add new classes if we think we need them.

side question: How much Object Oriented Programming do we want overall?

**Comment by sverhoeven** [comment]
About spectrum:
* Dont think Spectrum class should have a `load()` method, as it gets filled by [load_from_file](https://github.com/matchms/matchms/issues/51), having 2 load methods in different places is confusing.
* Filename property of a Spectrum object can be empty due to constructing it not from a file.
* The other properties in a Spectrum object will be things like precursor mass, parent mass, smiles, inchikey, charge. We have to decide if they should be accessed as `spectrum.precursor_mass` or `spectrum.meta.precursor_mass`, I am for less dots.

About Scores
* dont like calculate method. What it does is not clear from the name.

**Comment by florian-huber** [comment]
Sounds good @sverhoeven.
The Spectrum class has ancient historic roots in a former code... so all that is really needed is peaks, peak intensities and metadata. Not all metadata is needed, but we should at least keep everything that typically is expected in an MGF file.

**Comment by florian-huber** [comment]
@jspaaks, thanks a lot for digging through the code and coming up with the new proposed structure. Here some ideas on this from my side:

- I agree that the Spectrum() object can be much simpler and without its loading method etc.
- Still, to me it makes more sense to keep harmonization and filtering close to the importing step. This will also allow to perform additional analysis, tests, etc. on imported + cleaned data.

Would the following make sense to you as well?
```python
from matchms.importing import load_from_file, spectra_processing
from matchms.harmonization import harmfun1, harmfun2
from matchms.filtering import filter1
from matchms.similarity import simfun1, simfun2, simfun3
from spec2vec import spec2vec_similarity
from matchms import Scores

# Import and process spectra and reference_spectra
spectra = load_from_file(spectra_mgf)
spectra = spectra_processing(spectra, harmfun1, harmfun2, filter1)
reference_spectra = load_mgf_file(reference_spectra_mgf)
reference_spectra = spectra_processing(reference_spectra, harmfun1, harmfun2, filter1)

# Compute similarities
scores = Scores(spectra=spectra, reference_spectra=reference_spectra)
scores.add_similarity_function(simfun1, simfun2, simfun3, spec2vec_similarity) 
scores.calculate()

# Optional: export processed spectra
from matchms.exporting import save_spectra
save_spectra(reference_spectra, 'filename.mgf')
```

**Comment by florian-huber** [comment]
Thanks @jspaaks for digging through the code and coming up with the new proposed structure!
Here some ideas on that from my side: 

- I agree that the Spectrum() class can be without all the loading methods etc.
- Still, to me it seems better to keep the harmonization and filtering closer to the importing part. That would make it easier to run additional analysis, tests, etc. right after importing and processing the data.

Would such a structure make sense for you as well?

```python
from matchms.importing import load_mgf_file, spectra_processing
from matchms.harmonization import harmfun1, harmfun2 
from matchms.filtering import filter1
from matchms.similarity import simfun1, simfun2, simfun3
from spec2vec import spec2vec_similarity
from matchms import Scores

# Import and process spectra and reference spectra
spectra = load_mgf_file(spectra_mgf)
spectra = spectra_processing(spectra, harmfun1, harmfun2 , filter1)
reference_spectra = load_mgf_file(reference_spectra_mgf)
reference_spectra = spectra_processing(reference_spectra, harmfun1, harmfun2 , filter1)

# Compute similarities
scores = Scores(spectrum=spectra, reference_spectra=reference_spectra)
scores.add_similarity_function(simfun1, simfun2, simfun3, simfun4) 
scores.calculate()

# Optional: export processed spectra
from matchms.exporting import save_spectra
save_spectra(reference_spectra, 'filename.mgf')
```

**Comment by florian-huber** [comment]
In that structure there would be two classes: 
- ``Spectrum()`` -- containing: peaks, peak intensities, metadata
--> reduced version of current Spectrum() class from _ms_functions.py_
- ``Scores()`` -- containing functions to compute similarities and query/sort results
--> based on  current SimilarityMeasures() class from _similarity_measures.py_ as well as functionality from _ms_library_search.py_.

Further the functions:
- ``load_from_file`` -- input: mgf file, output: list of ``Spectrum()`` objects
- ``spectra_processing`` -- input:  ``Spectrum()`` object(s), output: ``Spectrum()`` object(s)
--> currently both importing + filtering are called from load_MGF_ function from _ms_functions.py_
--> harmonization is currently only partly in the main code and to some extend still only exists in form of notebooks.
- ``save_spectra`` -- input: ``Spectrum()`` object(s), output: MGF file (optional: other file formats as well).

**Comment by florian-huber** [comment]
An alternative choice could be to work with a larger ``Spectra()`` class instead to avoid handling arrays or lists of ``Spectrum()`` objects. I don't have a clear opinion on what's better...

```python
from matchms.harmonization import harmfun1, harmfun2
from matchms.filtering import filter1
from matchms.similarity import simfun1, simfun2, simfun3
from spec2vec import spec2vec_similarity
from matchms import Spectra, Scores

# Import and process spectra and reference_spectra
spectra = Spectra()
spectra.load_from_file(spectra_mgf)
spectra.processing(harmfun1, harmfun2, filter1)
reference_spectra = Spectra()
reference_spectra.load_from_file(reference_spectra_mgf)
reference_spectra.processing(harmfun1, harmfun2, filter1)

# Compute similarities
scores = Scores(spectra=spectra, reference_spectra=reference_spectra)
scores.add_similarity_function(simfun1, simfun2, simfun3, spec2vec_similarity) 
scores.calculate()

# Optional: export processed spectra
reference_spectra.save_spectra('filename.mgf')
```

**Comment by jspaaks** [comment]
I would need to think a bit more about the details later, but here is a quick response to https://github.com/matchms/matchms/issues/41#issuecomment-606674110 at least. Keep in mind these are not facts/written in stone, just my opinions

```python
from matchms.importing import load_from_file, spectra_processing
from matchms.harmonization import harmfun1, harmfun2
from matchms.filtering import filter1
```

- ``spectra_processing`` should not be in a module called ``importing``. The name ``importing`` suggests that when I have some information on disk, the functionality from ``importing`` can help me get that information into memory (without any changes).
- all the harmonization functions are filters, so you could opt to rename ``matchms.harmonization`` to ``matchms.filtering``. I went with ``harmonization`` because it has a narrower definition than ``filtering``. But for the purpose of demonstration, that means that the above code block could be made better by replacing it with 

    ```python
    from matchms.importing import load_from_file 
    from matchms.filtering import filter1, harmfun1, harmfun2, spectra_processing
    ```
- ``spectra_processing`` should be renamed to be more specific as well --currently what a person can get from this name is "it does something with spectra".


As to introducing the larger ``Spectra`` class (https://github.com/matchms/matchms/issues/41#issuecomment-606702327), I don't think it is so different from the original proposal, although I don't see why ``Spectra`` needs a ``.processing()``. Also, by exposing ``.processing`` in this way, Spectra no longer has the meaning of "an in-memory representation of what was on disk".

From a conceptual point of view, the introduction of a Spectra does not help (me) reason about the code. Imagine you're telling somebody about what your code does. At no point would you (accurately) say "_and then I take the spectra and do X_". This is in contrast to having a Spectrum and a Scores: for example, I want to ``plot`` a Spectrum or filter out noise from a ``spectrum``, or fit a distribution to a ``spectrum``, or I want to compare it to another ``spectrum``. Similarly, I want to ``calculate`` the score values, or I want to ``__add__`` two parts of the scores table that perhaps were calculated on different machines for performance. 

I think @sverhoeven was a proponent of renaming ``Scores`` to something like ``ScoresBuilder`` or ``ScoresConfig`` or something to that effect, and of ``calculate`` being its own unbound function, such that you can do
```python
from matchms import Scores
from matchms.filters import filter1, filter2 
from matchms.similarity_functions import simfun1, simfun2 

filters = [filter1, filter2]
similarity_functions = [simfun1, simfun2]

config = ScoresConfig(spectrums=spectrums, reference_spectrums=reference_spectrums,
                      filters=filters, similarity_functions=similarity_functions)

scores = calculate_scores(config)
```

which may indeed be cleaner.


(after more thinking)

although that would cut the link between a row in ``scores`` and a ``reference_spectrum``, which would make it more difficult and more error-prone to do hierarchical analyses such as https://github.com/matchms/matchms/issues/38#issuecomment-605896633.

**Comment by jspaaks** [comment]
We talked about this already but for the record: I agree with Stefan in https://github.com/matchms/matchms/issues/41#issuecomment-606036418 that ``Spectrum`` should not have a ``.load``; instead ``load_from_file`` or ``load_from_mgf`` should return a ``Spectrum[]``

**Comment by florian-huber** [comment]
Thanks @jspaaks for clarifying this!

### 1. On the Importing & Processing part
Looks to me like that is conceptually easier to 'solve' and hence a good starting point for the refactoring?

I would agree with you that combining filtering and harmonization makes more sense, and that it shouldn't belong to the importing part! To avoid that ``spectra_processing`` sounds to vague one could maybe go for ``spectra_processing(spectra, harmonizer = [harmfun1,...], filters = [filter1, ...])`` , or spliting into two functions: ``spectra_harmonizing()`` and ``spectra_filtering()``, or ``filter_metadata()`` and ``filter_spectra()`` ?

One remark on the Spectrum vs Spectra class: We currently do 2 conceptually different types of filtering:
1. Filter individual spectra, e.g. remove low intensity peaks. 
2. Remove entire spectra for the list/array/set of spectra, e.g. when they don't match certain quality criteria. 
Still, it sounds like going for a ``Spectrum()`` class makes more sense.

```python
from matchms.importing import load_from_mgf
from matchms.harmonizing import spectra_harmonizing, harmfun1, harmfun2
from matchms.filtering import spectra_filtering, filter1, filter2

# Import and process spectra and reference_spectra
spectra = load_from_file(spectra_mgf)
spectra = spectra_harmonizing(spectra, harmfun1, harmfun2)
spectra = spectra_filtering(spectra, filter1, filter2)
reference_spectra = load_from_file(reference_spectra_mgf)
reference_spectra = spectra_harmonizing(reference_spectra, harmfun1, harmfun2)
reference_spectra = spectra_filtering(reference_spectra, filter1, filter2)

# Optional: export processed spectra
from matchms.exporting import save_spectra
save_spectra(reference_spectra, 'filename.mgf')
```

**Comment by florian-huber** [comment]
### 2. Applying the different similarity measures to a set of spectra

- It still makes more sense to me to keep the filters out of this step and rather make it part of the workflow until it comes to this point.
- Two functionalities that we currently rely on are:
1. Calculate similarities (of desired type) between all pairs of spectra1 and spectra2. In principle spectra1 and spectra2 can be sets of spectra of any size.
--> Here a table is not the desired output format (and in many cases very inefficient). One could use a separate function ``calculate_score_matrix()`` that outputs a matrix of similarity values?
2. Query of (usually) a small number of spectra against a large set of reference spectra.
--> This is more complicated since it often relies on hierarchical processing (select the closest matches based on precursor_mass, then calculate the scores etc.). It could be done in steps as @jspaaks mentioned in #38. And here, a table output format is more desirable than for (1.)

The following would be a case where no ``Scores`` class is needed.
```python
from matchms.similarity_functions import simfun1, simfun2
from matchms.similarity_functions import calculate_score_matrix, calculate_score_table

scores_matrix = calculate_score_matrix(spectra=spectra, reference_spectra=reference_spectra,
                      similarity_function=simfun1)

score_table = calculate_score_table(spectra=spectra, reference_spectra=reference_spectra,
                      similarity_functions=[simfun1, simfun2])
```
Or using a ``Spectra_library()`` instead of a ``Scores`` class.
```python
from matchms import Spectra_library
from matchms.similarity_functions import simfun1, simfun2
from spec2vec import spec2vec_similarity
from matchms.similarity_functions import calculate_scores

# Calculating desired similarities
scores_matrix = calculate_scores(spectra=spectra, reference_spectra=reference_spectra,
                      similarity_function=simfun1)

# Library matching
ms_library = Spectra_library(spectra=reference_spectra)
found_matches = ms_library.match_spectra(query_spectra=spectra, 
                      similarity_functions=[simfun1, spec2vec_similarity])

```

or following your above example:
```python
from matchms import ScoresConfig
from matchms.similarity_functions import calculate_scores, simfun1, simfun2 

config_matrix = ScoresConfig(spectra=spectra, reference_spectra=reference_spectra,
                      similarity_functions=[simfun1], outputstyle='matrix')
scores_matrix = calculate_scores(config_matrix)

config_table = ScoresConfig(spectra=spectra, reference_spectra=reference_spectra,
                      similarity_functions=[simfun1,` simfun2], outputstyle='table')
scores_table = calculate_scores(config_table)
```

**Comment by jspaaks** [comment]
(continuation of https://github.com/matchms/matchms/issues/41#issuecomment-606795954)

after some more thinking:

```python
from matchms import calculate_scores
from matchms.filters import filter1, filter2 
from matchms.similarity import simfun1, simfun2 

filters = [filter1, filter2]
similarity_functions = [simfun1, simfun2]

# scores is an instance of Scores
scores = calculate_scores(spectrum=spectrum,
                          reference_spectrums=reference_spectrums,
                          filters=filters,
                          similarity_functions=similarity_functions)  
```
This would facilitate easy handling of scores later, like:
```python
scores_subset = scores.sort_by("cosine").top(1000)
```
without accidentally losing 
- the relation between ``reference_spectrums`` and the corresponding similarity scores
- the relation between measured/query spectrum(s) and the corresponding scores
- how the ``reference_spectrums`` were filtered
- how ``spectrum`` was filtered (harmonized)
- what similarity functions were applied

BTW this structure is already able to facilitate having multiple measured/query spectrums, the code snippet immediately above would not need to change (except ``calculate_scores``'s ``spectrum`` argument would become ``spectrums``. 


For me, this issue is now resolved: we know we need:

1. a ``Spectrum`` class (I've created https://github.com/matchms/matchms/issues/63)
1. a ``Scores`` class (I've created https://github.com/matchms/matchms/issues/64)
1. a ``calculate_scores`` function (I've created https://github.com/matchms/matchms/issues/65)

I propose that we postpone the discussion on https://github.com/matchms/matchms/issues/41#issuecomment-606849731 until after we have a first implementation. That way, we can gain some experience with the proposed implementation, see what works and what doesn't, etc. To me, the biggest job is to refactor the content of the existing functions into their new shape --renaming a function, moving it around within matchms, or changing a function from bound to unbound is easy in comparison. 

Let me you know what y'all think!

**Comment by jspaaks** [comment]
This is part of https://github.com/matchms/matchms/pull/67. Once that is merged, this issue should be closed.
